kernel/arch/dreamcast/hardware/sq.c

```diff
@@ -2,165 +2,180 @@
-    n >>= 5;
-
-    while(n--) {
-        __asm__("pref @%0" : : "r"(d));
-        d += 8;
-    }
+    return sq_set32(dest, c, n);
+}
 
-    /* Wait for both store queues to complete */
-    d = (unsigned int *)MEM_AREA_SQ_BASE;
-    d[0] = d[8] = 0;
+/* Fills n bytes at dest with short c, dest must be 32-byte aligned */
+void * sq_set16(void *dest, uint32_t c, int n) {
+    /* Duplicate low 16-bits of c into high 16-bits */
+    c = c & 0xffff;
+    c = (c << 16) | c;
 
-    return s;
+    return sq_set32(dest, c, n);
 }
 
-/* fills n bytes at s with short c, s must be 32-byte aligned */
-void * sq_set16(void *s, uint32 c, int n) {
-    unsigned int *d = (unsigned int *)(void *)
-                      (MEM_AREA_SQ_BASE | (((unsigned long)s) & 0x03ffffe0));
+/* Fills n bytes at dest with int c, dest must be 32-byte aligned */
+void * sq_set32(void *dest, uint32_t c, int n) {
+    uint32_t *d = SQ_MASK_DEST(dest);
 
     /* Set store queue memory area as desired */
-    QACR0 = ((((unsigned int)s) >> 26) << 2) & 0x1c;
-    QACR1 = ((((unsigned int)s) >> 26) << 2) & 0x1c;
-
-    /* duplicate low 16-bits of c into high 16-bits */
-    c = c & 0xffff;
-    c = (c << 16) | c;
+    SET_QACR_REGS(dest);
 
     /* Fill both store queues with c */
     d[0] = d[1] = d[2] = d[3] = d[4] = d[5] = d[6] = d[7] =
-        d[8] = d[9] = d[10] = d[11] = d[12] = d[13] = d[14] = d[15] = c;
+            d[8] = d[9] = d[10] = d[11] = d[12] = d[13] = d[14] = d[15] = c;
 
     /* Write them as many times necessary */
     n >>= 5;
 
     while(n--) {
-        __asm__("pref @%0" : : "r"(d));
+        __builtin_prefetch(d);
         d += 8;
     }
 
     /* Wait for both store queues to complete */
-    d = (unsigned int *)MEM_AREA_SQ_BASE;
+    d = (uint32_t *)MEM_AREA_SQ_BASE;
     d[0] = d[8] = 0;
 
-    return s;
+    return dest;
 }
 
-/* fills n bytes at s with int c, s must be 32-byte aligned */
-void * sq_set32(void *s, uint32 c, int n) {
-    unsigned int *d = (unsigned int *)(void *)
-                      (MEM_AREA_SQ_BASE | (((unsigned long)s) & 0x03ffffe0));
+/* Clears n bytes at dest, dest must be 32-byte aligned */
+void sq_clr(void *dest, int n) {
+    sq_set32(dest, 0, n);
+}
 
-    /* Set store queue memory area as desired */
-    QACR0 = ((((unsigned int)s) >> 26) << 2) & 0x1c;
-    QACR1 = ((((unsigned int)s) >> 26) << 2) & 0x1c;
+#define PVR_LMMODE    (*(volatile uint32_t *)(void *)0xa05f6884)
+#define PVR_DMA_DEST  (*(volatile uint32_t *)(void *)0xa05f6808)
 
-    /* Fill both store queues with c */
-    d[0] = d[1] = d[2] = d[3] = d[4] = d[5] = d[6] = d[7] =
-        d[8] = d[9] = d[10] = d[11] = d[12] = d[13] = d[14] = d[15] = c;
+/* Copies n bytes from src to dest (in VRAM), dest must be 32-byte aligned */
+void * sq_cpy_pvr(void *dest, const void *src, int n) {
+    if(PVR_DMA_DEST != 0) {
+        dbglog(DBG_ERROR, "sq_cpy_pvr: Previous DMA has not finished\n");
+        errno = EINPROGRESS;
+        return dest;
```

Comment:
```suggestion
        return NULL;
```

I think just returning a NULL pointer here and not manipulating errno is a better idea.

Reply:
Sounds good.  I just used the same approach from here: https://github.com/KallistiOS/KallistiOS/blob/1526b24db29723634ae096a139f70a918eb428fd/kernel/arch/dreamcast/hardware/pvr/pvr_dma.c#L88